File: Class 17/Studio Solution/Rocket.ts
import { Astronaut } from "./Astronaut";
import { Cargo } from "./Cargo";
import { Payload } from "./Payload";

export class Rocket {
  name: string;
  totalCapacityKg: number;
  cargoItems: Cargo[] = [];
  astronauts: Astronaut[] = [];

  constructor(name: string, totalCapacityKg: number) {
    this.name = name;
    this.totalCapacityKg = totalCapacityKg;
  }

  sumMass(items: Payload[]): number {
    let sum: number = 0
    if(items.length > 0) {
      sum = items.map(i => i.massKg).reduce((acc, i) => acc + i)
    }
    return sum
  }

  currentMassKg(): number {
    return this.sumMass(this.astronauts) + this.sumMass(this.cargoItems)
  }

  canAdd(item: Payload): boolean {
    return this.currentMassKg() + item.massKg <= this.totalCapacityKg
  }

  addCargo(cargo: Cargo): boolean {
    if(this.canAdd(cargo)) {
      this.cargoItems.push(cargo)
      return true
    } else {
      return false;
    }
  }

  addAstronaut(astronaut: Astronaut): boolean {
    if(this.canAdd(astronaut)) {
      this.astronauts.push(astronaut)
      return true
    } else {
      return false;
    }
  }
}

File: Class 17/Studio Demo/Models/Rocket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Studio_Demo.Models
{
    public class Rocket
    {
        public string Name { get; set; }
        public int TotalCapacityKg { get; set; }
        public List<Cargo> CargoItems { get; set; }
        public List<Astronaut> Astronauts { get; set; }

        public Rocket(string name, int totalCapacityKg)
        {
            this.Name = name;
            this.TotalCapacityKg = totalCapacityKg;
            this.CargoItems = new List<Cargo>();
            this.Astronauts = new List<Astronaut>();
        }

        public double SumMass(List<Payload> items)
        {
            double sum = 0;
            if(items.Count > 0)
            {
                sum = items.Sum(i => i.MassKg);
            }
            return sum;
        }

        public double CurrentMassKg()
        {
            var cargoList = CargoItems.ToList<Payload>();
            var astronautList = Astronauts.ToList<Payload>();
            return SumMass(cargoList) + SumMass(astronautList);
        }

        public bool CanAdd(Payload item)
        {
            return CurrentMassKg() + item.MassKg <= this.TotalCapacityKg;
        }

        public bool AddCargo(Cargo cargo)
        {
            if(CanAdd(cargo))
            {
                CargoItems.Add(cargo);
                return true;
            }
            return false;
        }

        public bool AddAstronaut(Astronaut astronaut)
        {
            if(CanAdd(astronaut))
            {
                Astronauts.Add(astronaut);
                return true;
            }
            return false;
        }
    }
}
